mem/buffers.go

```diff
@@ -0,0 +1,149 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package mem provides utilities that facilitate memory reuse in byte slices
+// that are used as buffers.
+//
+// # Experimental
+//
+// Notice: All APIs in this package are EXPERIMENTAL and may be changed or
+// removed in a later release.
+package mem
+
+import (
+	"fmt"
+	"sync/atomic"
+)
+
+// A Buffer represents a reference counted piece of data (in bytes) that can be
+// acquired by a call to NewBuffer() or Copy(). A reference to a Buffer may be
+// released by calling Free(), which invokes the given free function only after
+// all references are released.
+//
+// Note that a Buffer is not safe for concurrent access and instead each
+// goroutine should use its own reference to the data, which can be acquired via
+// a call to Ref().
+//
+// Attempts to access the underlying data after releasing the reference to the
+// Buffer will panic.
+type Buffer struct {
+	data  []byte
+	refs  *atomic.Int32
+	free  func()
+	freed bool
+}
+
+// NewBuffer creates a new Buffer from the given data, initializing the
+// reference counter to 1. The given free function is called when all references
+// to the returned Buffer are released.
+//
+// Note that the backing array of the given data is not copied.
+func NewBuffer(data []byte, free func([]byte)) *Buffer {
+	b := &Buffer{data: data, refs: new(atomic.Int32)}
+	if free != nil {
+		b.free = func() { free(data) }
+	}
```

Comment:
This is actually important because the `Split` function changes `b.data`, so capturing this here avoids potential pitfalls down the line. It was initially set up that way but became too complicated to manage